engine/function/src/templates/Numeric.ftl

```diff
@@ -1478,84 +1478,141 @@ public class Numeric {
+    public static double sum(${pt.vector} values) {
         if (values == null) {
-            return ${pt.null};
+            return NULL_DOUBLE;
         }
 
         double sum = 0;
 
         try ( final ${pt.vectorIterator} vi = values.iterator() ) {
             while ( vi.hasNext() ) {
                 final ${pt.primitive} c = vi.${pt.iteratorNext}();
-    <#if pt.valueType.isFloat >
-                if (isNaN(c)) {
-                    return ${pt.boxed}.NaN;
+
+                if (isNaN(c) || isNaN(sum)) {
+                    return Double.NaN;
                 }
-    </#if>
+
                 if (!isNull(c)) {
                     sum += c;
                 }
             }
         }
 
-        return (${pt.primitive}) (sum);
+        return sum;
     }
+    <#else>
+    public static long sum(${pt.vector} values) {
+        if (values == null) {
+            return NULL_LONG;
+        }
+
+        long sum = 0;
+
+        try ( final ${pt.vectorIterator} vi = values.iterator() ) {
+            while ( vi.hasNext() ) {
+                final ${pt.primitive} c = vi.${pt.iteratorNext}();
+
+                if (!isNull(c)) {
+                    sum += c;
+                }
+            }
+        }
+
+        return sum;
+    }
+    </#if>
 
     /**
      * Returns the sum.  Null values are excluded.
      *
      * @param values values.
      * @return sum of non-null values.
      */
-    public static ${pt.primitive} sum(${pt.primitive}... values) {
+    <#if pt.valueType.isFloat >
+    public static double sum(${pt.primitive}... values) {
         if (values == null) {
-            return ${pt.null};
+            return NULL_DOUBLE;
         }
 
         return sum(new ${pt.vectorDirect}(values));
     }
+    <#else>
+    public static long sum(${pt.primitive}... values) {
+        if (values == null) {
+            return NULL_LONG;
+        }
+
+        return sum(new ${pt.vectorDirect}(values));
+    }
+    </#if>
 
     /**
      * Returns the product.  Null values are excluded.
      *
      * @param values values.
      * @return product of non-null values.
      */
-    public static ${pt.primitive} product(${pt.vector} values) {
+    <#if pt.valueType.isFloat >
+    public static double product(${pt.vector} values) {
         if (values == null) {
-            return ${pt.null};
+            return NULL_DOUBLE;
         }
 
-        ${pt.primitive} prod = 1;
+        double prod = 1;
         int count = 0;
-    <#if pt.valueType.isFloat >
-        long zeroCount = 0;
-        long infCount = 0;
-    </#if>
+        boolean hasZero = false;
+        boolean hasInf = false;
 
         try ( final ${pt.vectorIterator} vi = values.iterator() ) {
             while ( vi.hasNext() ) {
                 final ${pt.primitive} c = vi.${pt.iteratorNext}();
-    <#if pt.valueType.isFloat >
-                if (isNaN(c)) {
-                    return ${pt.boxed}.NaN;
+
+                if (isNaN(c) || isNaN(prod)) {
```

Comment:
Also here. I don't see how the algorithm progresses to the point where `prod` contains NaN.

Reply:
You're right on this one.  I did hasty copying.